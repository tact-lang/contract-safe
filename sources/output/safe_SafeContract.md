# TACT Compilation Report
Contract: SafeContract
BOC Size: 1926 bytes

# Types
Total Types: 21

## StateInit
TLB: `_ code:^cell data:^cell = StateInit`
Signature: `StateInit{code:^cell,data:^cell}`

## Context
TLB: `_ bounced:bool sender:address value:int257 raw:^slice = Context`
Signature: `Context{bounced:bool,sender:address,value:int257,raw:^slice}`

## SendParameters
TLB: `_ bounce:bool to:address value:int257 mode:int257 body:Maybe ^cell code:Maybe ^cell data:Maybe ^cell = SendParameters`
Signature: `SendParameters{bounce:bool,to:address,value:int257,mode:int257,body:Maybe ^cell,code:Maybe ^cell,data:Maybe ^cell}`

## SafeDeployParams
TLB: `_ deploy:coins devFee:coins = SafeDeployParams`
Signature: `SafeDeployParams{deploy:coins,devFee:coins}`

## Deploy
TLB: `deploy#946a98b6 queryId:uint64 = Deploy`
Signature: `Deploy{queryId:uint64}`

## DeployOk
TLB: `deploy_ok#aff90f57 queryId:uint64 = DeployOk`
Signature: `DeployOk{queryId:uint64}`

## ChangeOwner
TLB: `change_owner#0f474d03 newOwner:address = ChangeOwner`
Signature: `ChangeOwner{newOwner:address}`

## SafeParameters
TLB: `_ timeout:uint32 requestPrice:coins = SafeParameters`
Signature: `SafeParameters{timeout:uint32,requestPrice:coins}`

## SafeOperations
TLB: `_ ops:dict<int, ^SafeOperation{transfer:Maybe SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell},parameters:Maybe SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}}}> count:uint8 = SafeOperations`
Signature: `SafeOperations{ops:dict<int, ^SafeOperation{transfer:Maybe SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell},parameters:Maybe SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}}}>,count:uint8}`

## SafeOperation
TLB: `_ transfer:Maybe SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell} parameters:Maybe SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}} = SafeOperation`
Signature: `SafeOperation{transfer:Maybe SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell},parameters:Maybe SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}}}`

## SafeOperationTransfer
TLB: `_ to:address value:coins mode:uint8 bounce:bool body:Maybe ^cell = SafeOperationTransfer`
Signature: `SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell}`

## SafeOperationUpdateParameters
TLB: `_ parameters:SafeParameters{timeout:uint32,requestPrice:coins} = SafeOperationUpdateParameters`
Signature: `SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}}`

## EventSafeDeployed
TLB: `event_safe_deployed#51a47444 by:address address:address = EventSafeDeployed`
Signature: `EventSafeDeployed{by:address,address:address}`

## EventVote
TLB: `event_vote#9dc5d639 yes:bool address:address = EventVote`
Signature: `EventVote{yes:bool,address:address}`

## EventVoteCompleted
TLB: `event_vote_completed#e7e1407c success:bool = EventVoteCompleted`
Signature: `EventVoteCompleted{success:bool}`

## SafeDeployment
TLB: `safe_deployment#6eb08253 remaining:address = SafeDeployment`
Signature: `SafeDeployment{remaining:address}`

## SafeRequestOperation
TLB: `safe_request_operation#ff7de06c ops:SafeOperations{ops:dict<int, ^SafeOperation{transfer:Maybe SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell},parameters:Maybe SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}}}>,count:uint8} = SafeRequestOperation`
Signature: `SafeRequestOperation{ops:SafeOperations{ops:dict<int, ^SafeOperation{transfer:Maybe SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell},parameters:Maybe SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}}}>,count:uint8}}`

## VoteSuccess
TLB: `vote_success#cdf42430 args:VoteArgs{safe:address,owners:dict<address, bool>,ownersCount:uint32,treshold:uint32,timeout:uint64,ops:SafeOperations{ops:dict<int, ^SafeOperation{transfer:Maybe SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell},parameters:Maybe SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}}}>,count:uint8}} = VoteSuccess`
Signature: `VoteSuccess{args:VoteArgs{safe:address,owners:dict<address, bool>,ownersCount:uint32,treshold:uint32,timeout:uint64,ops:SafeOperations{ops:dict<int, ^SafeOperation{transfer:Maybe SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell},parameters:Maybe SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}}}>,count:uint8}}}`

## VoteFailure
TLB: `vote_failure#9895f337 args:VoteArgs{safe:address,owners:dict<address, bool>,ownersCount:uint32,treshold:uint32,timeout:uint64,ops:SafeOperations{ops:dict<int, ^SafeOperation{transfer:Maybe SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell},parameters:Maybe SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}}}>,count:uint8}} = VoteFailure`
Signature: `VoteFailure{args:VoteArgs{safe:address,owners:dict<address, bool>,ownersCount:uint32,treshold:uint32,timeout:uint64,ops:SafeOperations{ops:dict<int, ^SafeOperation{transfer:Maybe SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell},parameters:Maybe SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}}}>,count:uint8}}}`

## VoteArgs
TLB: `_ safe:address owners:dict<address, bool> ownersCount:uint32 treshold:uint32 timeout:uint64 ops:SafeOperations{ops:dict<int, ^SafeOperation{transfer:Maybe SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell},parameters:Maybe SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}}}>,count:uint8} = VoteArgs`
Signature: `VoteArgs{safe:address,owners:dict<address, bool>,ownersCount:uint32,treshold:uint32,timeout:uint64,ops:SafeOperations{ops:dict<int, ^SafeOperation{transfer:Maybe SafeOperationTransfer{to:address,value:coins,mode:uint8,bounce:bool,body:Maybe ^cell},parameters:Maybe SafeOperationUpdateParameters{parameters:SafeParameters{timeout:uint32,requestPrice:coins}}}>,count:uint8}}`

## VoteDeploy
TLB: `vote_deploy#56137615 queryId:uint64 = VoteDeploy`
Signature: `VoteDeploy{queryId:uint64}`

# Get Methods
Total Get Methods: 5

## parameters

## treshold

## seqno

## isOwner
Argument: address

## owners

# Error Codes
2: Stack undeflow
3: Stack overflow
4: Integer overflow
5: Integer out of expected range
6: Invalid opcode
7: Type check error
8: Cell overflow
9: Cell underflow
10: Dictionary error
13: Out of gas error
32: Method ID not found
34: Action is invalid or not supported
37: Not enough TON
38: Not enough extra-currencies
128: Null reference exception
129: Invalid serialization prefix
130: Invalid incoming message
131: Constraints error
132: Access denied
133: Contract stopped
134: Invalid argument
135: Code of a contract was not found
136: Invalid address
137: Masterchain support is not enabled for this contract
3094: Already signed or not an owner
4755: Timeout
5165: Not enough value
7464: Request price must be less or equals than 10 TON
8884: Wrong owners
17702: Not timeouted
25849: Not enough value to deploy a Safe
31822: Timeout must be less than a year
32453: Wrong safe address
36682: Sender is not safe
37479: Wrong treshold
38009: Request price must be more or equals than 1 TON
40810: Completed
44757: Exactly one operation must be specified
45379: Wrong signer address
46307: Not a member
62638: Wrong owners count