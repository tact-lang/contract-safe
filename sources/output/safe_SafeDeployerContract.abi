{"name":"SafeDeployerContract","types":[{"name":"StateInit","header":null,"fields":[{"name":"code","type":{"kind":"simple","type":"cell","optional":false}},{"name":"data","type":{"kind":"simple","type":"cell","optional":false}}]},{"name":"Context","header":null,"fields":[{"name":"bounced","type":{"kind":"simple","type":"bool","optional":false}},{"name":"sender","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"raw","type":{"kind":"simple","type":"slice","optional":false}}]},{"name":"SendParameters","header":null,"fields":[{"name":"bounce","type":{"kind":"simple","type":"bool","optional":false}},{"name":"to","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"mode","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"body","type":{"kind":"simple","type":"cell","optional":true}},{"name":"code","type":{"kind":"simple","type":"cell","optional":true}},{"name":"data","type":{"kind":"simple","type":"cell","optional":true}}]},{"name":"SafeDeployParams","header":null,"fields":[{"name":"deploy","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"devFee","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"Deploy","header":2490013878,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"DeployOk","header":2952335191,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"ChangeOwner","header":256331011,"fields":[{"name":"newOwner","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"SafeParameters","header":null,"fields":[{"name":"timeout","type":{"kind":"simple","type":"uint","optional":false,"format":32}},{"name":"requestPrice","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"timelock","type":{"kind":"simple","type":"uint","optional":false,"format":32}}]},{"name":"SafeOperations","header":null,"fields":[{"name":"ops","type":{"kind":"dict","key":"int","value":"SafeOperation","valueFormat":"ref"}},{"name":"count","type":{"kind":"simple","type":"uint","optional":false,"format":8}}]},{"name":"SafeOperation","header":null,"fields":[{"name":"transfer","type":{"kind":"simple","type":"SafeOperationTransfer","optional":true}},{"name":"parameters","type":{"kind":"simple","type":"SafeOperationUpdateParameters","optional":true}},{"name":"add","type":{"kind":"simple","type":"SafeOperationAdd","optional":true}},{"name":"remove","type":{"kind":"simple","type":"SafeOperationRemove","optional":true}},{"name":"replace","type":{"kind":"simple","type":"SafeOperationReplace","optional":true}}]},{"name":"SafeOperationAdd","header":null,"fields":[{"name":"owner","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"SafeOperationRemove","header":null,"fields":[{"name":"owner","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"SafeOperationReplace","header":null,"fields":[{"name":"old","type":{"kind":"simple","type":"address","optional":false}},{"name":"new","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"SafeOperationTransfer","header":null,"fields":[{"name":"to","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"mode","type":{"kind":"simple","type":"uint","optional":false,"format":8}},{"name":"bounce","type":{"kind":"simple","type":"bool","optional":false}},{"name":"body","type":{"kind":"simple","type":"cell","optional":true}}]},{"name":"SafeOperationUpdateParameters","header":null,"fields":[{"name":"parameters","type":{"kind":"simple","type":"SafeParameters","optional":false}}]},{"name":"EventSafeDeployed","header":1369732164,"fields":[{"name":"by","type":{"kind":"simple","type":"address","optional":false}},{"name":"address","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"EventVote","header":2646988345,"fields":[{"name":"yes","type":{"kind":"simple","type":"bool","optional":false}},{"name":"address","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"EventVoteCompleted","header":3890299004,"fields":[{"name":"success","type":{"kind":"simple","type":"bool","optional":false}}]},{"name":"SafeDeployment","header":1857061459,"fields":[{"name":"remaining","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"SafeRequestOperation","header":881562511,"fields":[{"name":"ops","type":{"kind":"simple","type":"SafeOperations","optional":false}}]},{"name":"VoteSuccess","header":2590601485,"fields":[{"name":"args","type":{"kind":"simple","type":"VoteArgs","optional":false}}]},{"name":"VoteFailure","header":415108203,"fields":[{"name":"args","type":{"kind":"simple","type":"VoteArgs","optional":false}}]},{"name":"VoteArgs","header":null,"fields":[{"name":"safe","type":{"kind":"simple","type":"address","optional":false}},{"name":"owners","type":{"kind":"dict","key":"address","value":"bool"}},{"name":"ownersCount","type":{"kind":"simple","type":"uint","optional":false,"format":32}},{"name":"treshold","type":{"kind":"simple","type":"uint","optional":false,"format":32}},{"name":"timeout","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"ops","type":{"kind":"simple","type":"SafeOperations","optional":false}}]},{"name":"VoteDeploy","header":1444115989,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]}],"receivers":[{"receiver":"internal","message":{"kind":"text","text":"Deploy new Safe"}},{"receiver":"internal","message":{"kind":"typed","type":"ChangeOwner"}},{"receiver":"internal","message":{"kind":"typed","type":"Deploy"}}],"getters":[{"name":"params","arguments":[],"returnType":{"kind":"simple","type":"SafeDeployParams","optional":false}},{"name":"owner","arguments":[],"returnType":{"kind":"simple","type":"address","optional":false}}],"errors":{"2":{"message":"Stack undeflow"},"3":{"message":"Stack overflow"},"4":{"message":"Integer overflow"},"5":{"message":"Integer out of expected range"},"6":{"message":"Invalid opcode"},"7":{"message":"Type check error"},"8":{"message":"Cell overflow"},"9":{"message":"Cell underflow"},"10":{"message":"Dictionary error"},"13":{"message":"Out of gas error"},"32":{"message":"Method ID not found"},"34":{"message":"Action is invalid or not supported"},"37":{"message":"Not enough TON"},"38":{"message":"Not enough extra-currencies"},"128":{"message":"Null reference exception"},"129":{"message":"Invalid serialization prefix"},"130":{"message":"Invalid incoming message"},"131":{"message":"Constraints error"},"132":{"message":"Access denied"},"133":{"message":"Contract stopped"},"134":{"message":"Invalid argument"},"135":{"message":"Code of a contract was not found"},"136":{"message":"Invalid address"},"137":{"message":"Masterchain support is not enabled for this contract"},"3094":{"message":"Already signed or not an owner"},"4755":{"message":"Timeout"},"5165":{"message":"Not enough value"},"7464":{"message":"Request price must be less or equals than 10 TON"},"8884":{"message":"Wrong owners"},"17702":{"message":"Not timeouted"},"25849":{"message":"Not enough value to deploy a Safe"},"31822":{"message":"Timeout must be less than a year"},"32453":{"message":"Wrong safe address"},"36682":{"message":"Sender is not safe"},"37479":{"message":"Wrong treshold"},"38009":{"message":"Request price must be more or equals than 1 TON"},"40810":{"message":"Completed"},"44757":{"message":"Exactly one operation must be specified"},"45379":{"message":"Wrong signer address"},"46307":{"message":"Not a member"},"62638":{"message":"Wrong owners count"}},"interfaces":["org.ton.introspection.v0","org.ton.abi.ipfs.v0","org.ton.deploy.lazy.v0","org.ton.chain.any.v0","org.ton.ownable.transferable","org.ton.ownable"]}