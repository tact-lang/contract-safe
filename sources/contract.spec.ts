import { Address, Dictionary, toNano } from "ton";
import { ContractSystem } from "@tact-lang/emulator";
import { SafeDeployerContract } from "./output/safe_SafeDeployerContract";
import { SafeContract, SafeOperation } from "./output/safe_SafeContract";
import { VoteContract, VoteArgs } from "./output/safe_VoteContract";

describe("contract", () => {
  it("should deploy correctly", async () => {
    let system = await ContractSystem.create();
    let owner = system.treasure("owner");
    let contract = system.open(await SafeDeployerContract.fromInit(owner.address, toNano('1'), toNano('1')));
    let tracker = system.track(contract);
    system.name(contract, "deployer");

    await contract.send(owner, { value: toNano(1) }, { $$type: "Deploy", queryId: 0n });
    await system.run();
    expect(tracker.collect()).toMatchSnapshot();
  });

  it('should create request', async () => {
    let system = await ContractSystem.create();
    let owner = system.treasure("owner");
    let contract = system.open(await SafeContract.fromInit(owner.address, 0n));
    system.name(contract, 'safe');
    let tracker = system.track(contract);

    // Deploy
    await contract.send(owner, { value: toNano(1) }, { $$type: 'SafeDeployment', remaining: owner.address });
    await system.run();
    expect(tracker.collect()).toMatchSnapshot();

    // Create request
    let operations = Dictionary.empty<bigint, SafeOperation>();
    await contract.send(owner, { value: toNano(1) }, {
      $$type: 'SafeRequestOperation',
      ops: {
        $$type: 'SafeOperations',
        ops: operations,
        count: BigInt(operations.size)
      }
    });
    await system.run();
    expect(tracker.collect()).toMatchSnapshot();
  });

  it('should create vote', async () => {

    // Create system
    let now = 1000;
    let system = await ContractSystem.create({ now });
    let safe = system.treasure('safe');
    let owner1 = system.treasure("owner1");
    let owner2 = system.treasure("owner2");
    let unknown = system.treasure("unknown");

    // Deploy signer
    let args: VoteArgs = {
      $$type: 'VoteArgs',
      safe: safe.address,
      owners: Dictionary.empty<Address, boolean>()
        .set(owner1.address, true)
        .set(owner2.address, true),
      ownersCount: 2n,
      treshold: 2n,
      timeout: BigInt(now + 10000),
      ops: {
        $$type: 'SafeOperations',
        ops: Dictionary.empty<bigint, SafeOperation>(),
        count: 0n
      }
    };
    let contract = system.open(await VoteContract.fromInit(args));
    let tracker = system.track(contract);
    system.name(contract, 'signer');
    await contract.send(safe, { value: toNano(1) }, { $$type: 'VoteDeploy', queryId: 0n });
    await system.run();
    expect(tracker.collect()).toMatchSnapshot('deploy');

    // Vote
    await contract.send(owner1, { value: toNano(1) }, 'YES');
    await system.run();
    expect(tracker.collect()).toMatchSnapshot('vote');

    // Double Vote
    await contract.send(owner1, { value: toNano(1) }, 'YES');
    await system.run();
    expect(tracker.collect()).toMatchSnapshot('double vote');

    // Invalid Voter
    await contract.send(unknown, { value: toNano(1) }, 'YES');
    await system.run();
    expect(tracker.collect()).toMatchSnapshot('invalid vote');

    // Final Vote
    await contract.send(owner2, { value: toNano(1) }, 'YES');
    await system.run();
    expect(tracker.collect()).toMatchSnapshot('vote');
  });
});
