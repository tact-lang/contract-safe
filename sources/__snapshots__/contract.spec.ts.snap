// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contract should create request 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{6EB0825380011F6E292AB9FE5825731C5F568DB15E68F69707390F1976DA6A7F173916703B9_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "@safe",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 8138n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "empty",
            },
            "bounce": true,
            "from": "@safe",
            "to": "@treasure(owner)",
            "type": "internal",
            "value": 990862000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`contract should create request 2`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{FF7DE06C004_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "@safe",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 15981n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{561376150000000000000000}",
              "type": "cell",
            },
            "bounce": true,
            "from": "@safe",
            "to": "kQCQX1XVtXgDF3BTzwaWn8E8xCneU4h6JloA_swDRj6iqvTk",
            "type": "internal",
            "value": 965358000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`contract should create vote: deploy 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{561376150000000000000000}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(safe)",
          "to": "@signer",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 6230n,
      },
    ],
  },
]
`;

exports[`contract should create vote: double vote 1`] = `
[
  {
    "$seq": 2,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "YES",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(owner1)",
          "to": "@signer",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "failed",
        "errorCode": 3094,
        "errorMessage": "Already signed or not an owner",
      },
      {
        "$type": "sent-bounced",
        "message": {
          "body": {
            "cell": "x{FFFFFFFF00000000594553}",
            "type": "cell",
          },
          "bounce": false,
          "from": "@signer",
          "to": "@treasure(owner1)",
          "type": "internal",
          "value": 993614000n,
        },
      },
    ],
  },
]
`;

exports[`contract should create vote: invalid vote 1`] = `
[
  {
    "$seq": 3,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "YES",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(unknown)",
          "to": "@signer",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "failed",
        "errorCode": 3094,
        "errorMessage": "Already signed or not an owner",
      },
      {
        "$type": "sent-bounced",
        "message": {
          "body": {
            "cell": "x{FFFFFFFF00000000594553}",
            "type": "cell",
          },
          "bounce": false,
          "from": "@signer",
          "to": "@treasure(unknown)",
          "type": "internal",
          "value": 993614000n,
        },
      },
    ],
  },
]
`;

exports[`contract should create vote: vote 1`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "YES",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(owner1)",
          "to": "@signer",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 14931n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{9DC5D639C00AB084690786D4F70011428E4B6BC532BFB47076AC4855DC7DA8144B22F85364F}",
              "type": "cell",
            },
            "to": null,
            "type": "external-out",
          },
        ],
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Vote accepted",
              "type": "text",
            },
            "bounce": false,
            "from": "@signer",
            "to": "@treasure(owner1)",
            "type": "internal",
            "value": 983833000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`contract should create vote: vote 2`] = `
[
  {
    "$seq": 4,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "YES",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(owner2)",
          "to": "@signer",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 20550n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{9DC5D639C000221528D67B5DFE51431A4453FDEA766A6642F184A3A3C5C6F16CC89465AB1EC}",
              "type": "cell",
            },
            "to": null,
            "type": "external-out",
          },
        ],
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Vote accepted",
              "type": "text",
            },
            "bounce": false,
            "from": "@signer",
            "to": "@treasure(owner2)",
            "type": "internal",
            "value": 978214000n,
          },
        ],
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{E7E1407CC_}",
              "type": "cell",
            },
            "to": null,
            "type": "external-out",
          },
        ],
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{CDF42430800AB60B6F3F99430109C97E1BE903020AA7211D122BF269A5ECE534A6217AE119500000002000000020000000000002AF8004_}
 x{817002_}
  x{BFC110A946B3DAEFF28A18D2229FEF53B35332178C251D1E2E378B6644A32D58F66_}
  x{BFD58423483C36A7B8008A14725B5E2995FDA383B56242AEE3ED40A25917C29B27E_}",
              "type": "cell",
            },
            "bounce": false,
            "from": "@signer",
            "to": "@treasure(safe)",
            "type": "internal",
            "value": 987446000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`contract should deploy correctly 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{946A98B60000000000000000}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "@deployer",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 8274n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{AFF90F570000000000000000}",
              "type": "cell",
            },
            "bounce": true,
            "from": "@deployer",
            "to": "@treasure(owner)",
            "type": "internal",
            "value": 990530000n,
          },
        ],
      },
    ],
  },
]
`;
